[Clang] Diagnose defaulted assignment operator with incompatible object parameter

cor3ntin opened the pull request:
Per https://eel.is/c++draft/dcl.fct.def.default#2.2, the explicit object parameter of a defaulted special member function must be of the same type as the one of an equivalent implicitly defaulted function, ignoring references.

Fixes #69233

cor3ntin committed:
[Clang] Diagnose defaulted assignment operator with incompatible object parameter.

Per https://eel.is/c++draft/dcl.fct.def.default#2.2, the explicit object
parameter of a defaulted special member function must be of the same type
as the one of an equivalent implicitly defaulted function,
ignoring references.

Fixes #69233

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -7748,6 +7748,24 @@ bool Sema::CheckExplicitlyDefaultedSpecialMember(CXXMethodDecl *MD,
+      else {
+        Diag(MD->getLocation(),
+             diag::err_defaulted_special_member_explicit_object_mismatch)
+            << (CSM == CXXMoveAssignment) << RD;
```
What do you have in mind? the first parameter? I am not sure how useful it is though - we certainly don't do it for other warnings here.

tbaederr replied:
Nothing specific. In line 7709 above, we pass `MD->getSourceRange()`.

cor3ntin committed:
add source range

cor3ntin committed:
Add tests for move.

Note that if there is an invalid move constructor,
overload resolution will try to pick a copy constructor instead.

AaronBallman commented:
> Per https://eel.is/c++draft/dcl.fct.def.default#2.2, the explicit object parameter of a defaulted special member function must be of the same type as the one of an equivalent implicitly defaulted function, ignoring references.

I could use an explanation as to how they must be the same type when 2.2 says "...may be an explicit object member function whose explicit object parameter is of (possibly different) type “reference to C”, in which case..."

cor3ntin commented:
> > Per https://eel.is/c++draft/dcl.fct.def.default#2.2, the explicit object parameter of a defaulted special member function must be of the same type as the one of an equivalent implicitly defaulted function, ignoring references.
> 
> I could use an explanation as to how they must be the same type when 2.2 says "...may be an explicit object member function whose explicit object parameter is of (possibly different) type “reference to C”, in which case..."

That wording is terrible, but it's hard to phrase it better.
The type may be different but it is always a reference to `C` where C is the same type.

Ie it can be either `C&` or `C&&`. this is consistent with the fact that we allow ref-qualified functions.
There is some context here https://lists.isocpp.org/core/2023/07/14608.php

I had a chat with core on mattermost and there is agreement that the wording could be made clearer editorially.

AaronBallman commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -7748,6 +7748,24 @@ bool Sema::CheckExplicitlyDefaultedSpecialMember(CXXMethodDecl *MD,
         HadError = true;
       }
     }
+    // [C++23][dcl.fct.def.default]/p2.2
```
This requirement seems to be a general one but you've implemented it as being specific to the copy and move assignment operators; can you not have this same situation with other operators (like comparisons, etc)?

cor3ntin replied:
It applies to special member functions - Which excludes comparison. and constructors can't have an explicit object parameters

AaronBallman replied:
Oh! I was thinking that spaceship was also a special member function that can be defaulted, but you're right, that's... not special. Thank you!

AaronBallman commented:
> > > Per https://eel.is/c++draft/dcl.fct.def.default#2.2, the explicit object parameter of a defaulted special member function must be of the same type as the one of an equivalent implicitly defaulted function, ignoring references.
> > 
> > 
> > I could use an explanation as to how they must be the same type when 2.2 says "...may be an explicit object member function whose explicit object parameter is of (possibly different) type “reference to C”, in which case..."
> 
> That wording is terrible, but it's hard to phrase it better. The type may be different but it is always a reference to `C` where C is the same type.
> 
> Ie it can be either `C&` or `C&&`. this is consistent with the fact that we allow ref-qualified functions. There is some context here https://lists.isocpp.org/core/2023/07/14608.php
> 
> I had a chat with core on mattermost and there is agreement that the wording could be made clearer editorially.

Thank you for the explanation!

AaronBallman reviewed (approved):
LGTM though something seems to be amiss with precommit CI and so it might be nice to get a clean run of that before landing (I don't insist).
